src/main/java/com/epam/aidial/core/controller/AccessControlBaseController.java

```diff
@@ -21,37 +22,66 @@ public abstract class AccessControlBaseController {
+            return Future.succeededFuture();
         }
 
+        // we should take a real user bucket not provided from resource
+        String actualUserLocation = BlobStorageUtil.buildInitiatorBucket(context);
+        String actualUserBucket = proxy.getEncryptionService().encrypt(actualUserLocation);
+
         ResourceDescription resource;
         try {
             resource = ResourceDescription.fromEncoded(type, urlDecodedBucket, decryptedBucket, path);
         } catch (Exception ex) {
             String errorMessage = ex.getMessage() != null ? ex.getMessage() : DEFAULT_RESOURCE_ERROR_MESSAGE.formatted(path);
-            return context.respond(HttpStatus.BAD_REQUEST, errorMessage);
+            context.respond(HttpStatus.BAD_REQUEST, errorMessage);
+            return Future.succeededFuture();
         }
 
-        handle(resource);
-        return Future.succeededFuture();
+        Future<Void> permissionFuture = proxy.getVertx().executeBlocking(() -> {
+            boolean hasReadAccess = isSharedResource(resource, actualUserBucket, actualUserLocation);
+            boolean hasWriteAccess = hasWriteAccess(path, decryptedBucket);
+            boolean hasAccess = checkFullAccess ? hasWriteAccess : hasReadAccess || hasWriteAccess;
+
+            if (!hasAccess) {
+                throw new HttpException(HttpStatus.FORBIDDEN, "You don't have an access to the bucket " + bucket);
+            }
+
+            return null;
+        });
+
+        permissionFuture.andThen((handler) -> {
+            if (handler.succeeded()) {
+                handle(resource);
+            } else {
+                Throwable error = handler.cause();
+                if (error instanceof HttpException httpException) {
+                    context.respond(httpException.getStatus(), httpException.getMessage());
+                } else {
+                    context.respond(HttpStatus.INTERNAL_SERVER_ERROR, error.getMessage());
+                }
+            }
+        });
+
+        return permissionFuture;
     }
 
     protected abstract Future<?> handle(ResourceDescription resource);
 
-    protected boolean isSharedWithMe(ResourceType type, String bucket, String filePath) {
-        String url = type.getGroup() + BlobStorageUtil.PATH_SEPARATOR + bucket + BlobStorageUtil.PATH_SEPARATOR + filePath;
-        return context.getApiKeyData().getAttachedFiles().contains(url);
+    protected boolean isSharedResource(ResourceDescription resource, String userBucket, String userLocation) {
+        // some per-request API-keys may have access to the resources implicitly
+        boolean isAutoShared = context.getApiKeyData().getAttachedFiles().contains(resource.getUrl());
+        // resource was shared explicitly by share API
+        boolean isExplicitlyShared = (proxy.getResourceService() != null && proxy.getShareService().hasReadAccess(userBucket, userLocation, resource));
```

Comment:
It is always checked. But we want to check only if it is not my resource and not auto shared